src/matchers/toHaveQueryParam/helpers.ts

```diff
@@ -0,0 +1,77 @@
+import { decodeQueryParams, encodeQueryParams, QueryParamConfig } from 'serialize-query-params'
+import { parse } from 'query-string'
+import isEqual from 'lodash.isequal'
+
+import { CustomMatcherMessage } from '../message'
+
+export interface ExpectedQueryParam {
+  name: string;
+  type: QueryParamConfig<any, any>
+  value: any;
+}
+
+type EncodedValue = string | string[] | null | (string | null)[] | undefined
+
+export const validateArguments = ({
+  name,
+  type,
+  locationSearch,
+  encodedReceivedValue
+}: Pick<ExpectedQueryParam, 'type' | 'name'> & {
+  locationSearch: Location['search'],
+  encodedReceivedValue: EncodedValue,
+}) => {
+  if (!type) {
+    throw new CustomMatcherMessage({
+      pass: false,
+      message: () => 'Please, make sure to provide a type argument in order to serialize a query param value'
+    })
+  }
+
+  if (!encodedReceivedValue) {
+    throw new CustomMatcherMessage({
+      pass: false,
+      message: () => `${name} wasn't found in ${locationSearch}`
+    })
+  }
+}
+
+export const encodeValues = ({
+  type,
+  name,
+  value,
+  locationSearch
+}: ExpectedQueryParam & {
+  locationSearch: Location['search'],
+}) => {
+  const encodedReceivedValue = parse(locationSearch)[name]
+
+  return {
+    encodedExpectedValue: encodeQueryParams(
+      { [name]: type },
+      { [name]: value }
+    )[name],
+    encodedReceivedValue,
+    decodedReceivedValue: decodeQueryParams(
+      { [name]: type },
+      { [name]: encodedReceivedValue }
+    )[name]
+  }
+}
+
+export const validateResult = ({
+  name,
+  value,
+  encodedExpectedValue,
+  encodedReceivedValue,
+  decodedReceivedValue
+}: Pick<ExpectedQueryParam, 'name' | 'value'> & {
+  encodedReceivedValue: EncodedValue,
+  encodedExpectedValue: EncodedValue,
+  decodedReceivedValue: EncodedValue
```

Comment:
TypeScript 2.8 provides a generic called `ReturnType<T>` which enables us to access the return type of an function by passing as the generic argument